types/wlr_data_device_manager.c

```diff
@@ -0,0 +1,185 @@
+	if (!wlr_wl_data_source_create(client, version, id)) {
+		wlr_log(L_ERROR, "Failed to create wlr_wl_data_source");
+		wl_resource_post_no_memory(res);
+		return;
+	}
+}
+
+static void data_device_manager_get_data_device(struct wl_client *client,
+		struct wl_resource *res, uint32_t id, struct wl_resource *seat_res) {
+	struct wlr_data_device_manager *manager = wl_resource_get_user_data(res);
+	struct wlr_seat_handle *seat_handle = wl_resource_get_user_data(seat_res);
+	struct wlr_data_device *device;
+	if (!(device = seat_ensure_data_device(manager, seat_handle->wlr_seat))) {
+		wl_resource_post_no_memory(res);
+		return;
+	}
+
+	if (seat_handle->data_device) {
+		// TODO: implement resource lists for seat related handles
+		//   this is a protocol violation, see the todos in wlr_seat.c
+		wl_resource_destroy(seat_handle->data_device);
+	}
+
+	seat_handle->data_device = wl_resource_create(client,
+		&wl_data_device_interface, wl_resource_get_version(res), id);
+	if (!seat_handle->data_device) {
+		wlr_log(L_ERROR, "Failed to create wl_data_device resource");
+		wl_resource_post_no_memory(res);
+		return;
+	}
+
+	wl_resource_set_implementation(seat_handle->data_device, &data_device_impl,
+		seat_handle, data_device_destroy);
+}
+
+struct wl_data_device_manager_interface data_device_manager_impl = {
+	.create_data_source = data_device_manager_create_data_source,
+	.get_data_device = data_device_manager_get_data_device
+};
+
+static void data_device_manager_bind(struct wl_client *client, void *data,
+		uint32_t version, uint32_t id) {
+	struct wlr_data_device_manager *manager = data;
+	assert(client && manager);
+	if (version > 3) {
+		wlr_log(L_ERROR, "Client requested unsupported data_device_manager "
+			"version, disconnecting");
+		wl_client_destroy(client);
+		return;
+	}
+	struct wl_resource *resource = wl_resource_create(
+			client, &wl_data_device_manager_interface, version, id);
+	if (!resource) {
+		wlr_log(L_ERROR, "Failed to allocate wl_data_device_manager resource");
+		wl_client_post_no_memory(client);
+		return;
+	}
+	wl_resource_set_implementation(resource, &data_device_manager_impl,
+			manager, NULL);
+}
+
+struct wlr_data_device_manager *wlr_data_device_manager_create(struct wl_display *dpy) {
+	struct wlr_data_device_manager *manager = calloc(1, sizeof(*manager));
+	if (!manager) {
+		wlr_log(L_ERROR, "Failed to allocated wlr_data_device_manager");
+		return NULL;
+	}
+
+	manager->global = wl_global_create(dpy, &wl_data_device_manager_interface, 3,
+		manager, data_device_manager_bind);
+	if (!manager->global) {
+		wlr_log(L_ERROR, "Failed to create global for wlr_data_device_manager");
+		free(manager);
+		return NULL;
+	}
+
+	return manager;
+}
+
+void wlr_data_device_manager_destroy(struct wlr_data_device_manager *manager) {
+	if (!manager) {
+		return;
+	}
+
+	// TODO: destroy remaining resources? cancel current selection?
+	//  if this is called why there are still resources active we will
+	//  always get problems
```

Comment:
oh this todo is actually wrong, data_device_manager does not manage any selections anymore (was moved with multi seat support). Will make sure to move/fix it.